docker/daemon/daemon_dest.go

```diff
@@ -22,36 +17,49 @@ import (
 
 type daemonImageDestination struct {
 	ref daemonReference
-	// For talking to imageLoadGoroutine
+	img *dockerArchive.Image
+
+	// Related to the goroutine which runs the `docker load`.
 	goroutineCancel context.CancelFunc
 	statusChannel   <-chan error
-	writer          *io.PipeWriter
-	tar             *tar.Writer
-	// Other state
-	committed bool // writer has been closed
+	committed       bool
+
+	// Horrible hack.
+	pipeW *io.PipeWriter
+	pipeR *io.PipeReader
 }
 
 // newImageDestination returns a types.ImageDestination for the specified image reference.
-func newImageDestination(systemCtx *types.SystemContext, ref daemonReference) (types.ImageDestination, error) {
+func newImageDestination(systemCtx *types.SystemContext, ref daemonReference) (_ types.ImageDestination, err error) {
 	// FIXME: Do something with ref
 	c, err := client.NewClient(client.DefaultDockerHost, "1.22", nil, nil) // FIXME: overridable host
 	if err != nil {
 		return nil, fmt.Errorf("Error initializing docker engine client: %v", err)
 	}
 
-	reader, writer := io.Pipe()
+	// XXX: This is a bit of a dirty hack, but we have to have a way to return
+	//      an error from Read() when c.ImageLoad is reading. For some reason,
+	//      goroutineCancel doesn't actually work with the Docker engine-api.
```

Comment:
Yes, but the fact that the `engine-api` doesn't use `ctx.WithCancel` properly is the reason that `CloseWithError` is necessary (it wouldn't be necessary otherwise).